pageserver/src/object_repository.rs

```diff
@@ -679,8 +707,19 @@ impl ObjectTimeline {
     fn do_gc(&self, conf: &'static PageServerConf) -> Result<()> {
         loop {
             thread::sleep(conf.gc_period);
+            let descendant_timelines =
+                page_cache::get_repository().get_timeline_descendants(self.timelineid)?;
+            // Calculate minimal last LSN for all descendant timelines.
+            // TODO: it may be reasonable to let user specify PITR horizon for each timeline individually
```

Comment:
It doesn't make much sense to do Min(lsn) across all timelines. Each timeline can have a wildly different "current" lsn. We should compute separate 'last_lsn' for each timeline.

Reply:
Sorry, I do not completely understand you concern. Assume that we have main branch and two forks of it: B1 and B2:
```
LSN   main B1 B2
    1        |
    2        |
    3        | \
    4        |   \ 
    5        |     \
    6        |     |
    7        |     | \
    8        |     |   \
    9        |     *    \
   10        |           |
   11        *           |
   12                    |
   13                    *
```
So B1 is forked from main at LSN=3, B2 forked from B1 at LSN=7, last_lsn(main)=11, last_lsn(B1)=9, last_lsn(B2)=13.
Right now we have common PITR interval for all timelines, i.e. gc_period=10.
It means that in main branch we can cleanup versions with LSN<=1, in B2 with LSN<=3 and in B1 we can not remove any version.  Assume that we re in GC for timeline "main", This check with calculation of minimal last_valid_lsn for all descendant timelines is used to prevent deletion of version which still may be needed for B1.